patches/tock/05-usb-cancel.patch

```diff
@@ -0,0 +1,747 @@
++            out_buffer: Default::default(),
+             client: OptionalCell::empty(),
+             tx_packet: OptionalCell::empty(),
+             pending_in: Cell::new(false),
+@@ -187,7 +189,7 @@ impl<'a, 'b, C: hil::usb::UsbController<'a>> ClientCtapHID<'a, 'b, C> {
+     fn send_packet_to_client(&'a self) -> bool {
+         // Copy the packet into a buffer to send to the client.
+         let mut buf: [u8; 64] = [0; 64];
+-        for (i, x) in self.buffer.buf.iter().enumerate() {
++        for (i, x) in self.out_buffer.buf.iter().enumerate() {
+             buf[i] = x.get();
+         }
+ 
+@@ -220,11 +222,7 @@ impl<'a, 'b, C: hil::usb::UsbController<'a>> ClientCtapHID<'a, 'b, C> {
+ 
+     fn cancel_in_transaction(&'a self) -> bool {
+         self.tx_packet.take();
+-        let result = self.pending_in.take();
+-        if result {
+-            self.controller().endpoint_cancel_in(1);
+-        }
+-        result
++        self.pending_in.take()
+     }
+ 
+     fn cancel_out_transaction(&'a self) -> bool {
+@@ -243,7 +241,10 @@ impl<'a, 'b, C: hil::usb::UsbController<'a>> hil::usb::Client<'a> for ClientCtap
+         self.client_ctrl.enable();
+ 
+         // Set up the interrupt in-out endpoint
+-        self.controller().endpoint_set_buffer(1, &self.buffer.buf);
++        self.controller()
++            .endpoint_set_in_buffer(1, &self.in_buffer.buf);
++        self.controller()
++            .endpoint_set_out_buffer(1, &self.out_buffer.buf);
+         self.controller()
+             .endpoint_in_out_enable(TransferType::Interrupt, 1);
+     }
+@@ -293,7 +294,7 @@ impl<'a, 'b, C: hil::usb::UsbController<'a>> hil::usb::Client<'a> for ClientCtap
+                 }
+ 
+                 if let Some(packet) = self.tx_packet.take() {
+-                    let buf = &self.buffer.buf;
++                    let buf = &self.in_buffer.buf;
+                     for i in 0..64 {
+                         buf[i].set(packet[i]);
+                     }
+@@ -346,6 +347,12 @@ impl<'a, 'b, C: hil::usb::UsbController<'a>> hil::usb::Client<'a> for ClientCtap
+             panic!("Unexpected tx_packet while a packet was being transmitted.");
+         }
+         self.pending_in.set(false);
++
++        // Clear any pending packet on the receiving side.
++        // It's up to the client to handle the transmitted packet and decide if they want to
++        // receive another packet.
++        self.cancel_out_transaction();
++
+         // Notify the client
+         self.client.map(|client| client.packet_transmitted());
+     }
+diff --git a/chips/nrf52/src/usbd.rs b/chips/nrf52/src/usbd.rs
+index 8c1992cc..972871d0 100644
+--- a/chips/nrf52/src/usbd.rs
++++ b/chips/nrf52/src/usbd.rs
+@@ -623,7 +623,7 @@ pub enum UsbState {
+ pub enum EndpointState {
+     Disabled,
+     Ctrl(CtrlState),
+-    Bulk(TransferType, EndpointDirection, BulkState),
++    Bulk(TransferType, Option<BulkInState>, Option<BulkOutState>),
```

Comment:
Is there an invariant that at least one of the option is not none? If yes, please document.